ros/src/twist_controller/dbw_helper.py

```diff
@@ -0,0 +1,204 @@
+    plt.plot(x_0, y_0, 'bo', label='point')
+    plt.plot([x_0, X[0]], [y_0, X[1]], 'b-', label='shortest distance')
+    delta = 2.0
+    plt.plot([X[0], X[0] + delta],
+             [X[1], X[1] + (2.0 * coefficients[0] * X[0] +
+                            coefficients[1]) * delta],
+             'g-', label='tangent')
+
+    plt.axis('equal')
+    plt.xlabel('x')
+    plt.ylabel('y')
+    plt.legend(loc='best')
+    plt.show()
+
+
+def shift_and_rotate_waypoints(pose, waypoints, points_to_use=None):
+    """
+    From a pose object transfrom a series of waypoints so that
+    the origin is at the pose position and the orientation matches
+    the yaw of the pose
+
+    Args:
+        pose (object) : A pose object
+        waypoints (list) : A list of waypoint objects
+        points_to_use (int) : How many points to use (None => all)
+
+    Returns:
+        x_coords (list): The transformed x-coordinates of waypoints
+        y_coords (list): The transformed y-coordinates of waypoints
+    """
+    x_coords = []
+    y_coords = []
+
+    _, _, yaw = get_euler(pose)
+    originX = pose.position.x
+    originY = pose.position.y
+
+    if points_to_use is None:
+        points_to_use = len(waypoints)
+
+
+    for i in range(points_to_use):
+
+        shift_x = waypoints[i].pose.pose.position.x - originX
+        shift_y = waypoints[i].pose.pose.position.y - originY
+
+        x = shift_x * cos(0 - yaw) - shift_y * sin(0 - yaw)
+        y = shift_x * sin(0 - yaw) + shift_y * cos(0 - yaw)
+
+        x_coords.append(x)
+        y_coords.append(y)
+
+    return x_coords, y_coords
+
+
+def is_waypoint_behind(pose, waypoint):
+    """Take a waypoint and a pose , do a coordinate system transformation
+    setting the origin at the position of the pose object and as x-axis
+    the orientation of the z-axis of the pose
+
+    Args:
+        pose (object) : A pose object
+        waypoints (object) : A waypoint object
+
+    Returns:
+        bool : True if the waypoint is behind the car False if in front
+
+    """
+    _, _, yaw = get_euler(pose)
+    originX = pose.position.x
+    originY = pose.position.y
+
+    shift_x = waypoint.pose.pose.position.x - originX
+    shift_y = waypoint.pose.pose.position.y - originY
+
+    x = shift_x * cos(0 - yaw) - shift_y * sin(0 - yaw)
+
+    if x > 0:
+        return False
+    return True
+
+
+def cte(pose, waypoints):
+    """
+    From a pose object and a series of waypoints, calculate the distance
+    (cross track error) of the point, defined by the pose, to the path,
+    defined by the series of waypoints.
+
+    Args:
+        pose (object) : A pose object
+        waypoints (list) : A list of waypoint objects
+
+    Returns:
+        cte (float) : the cross track error (signed)
+    """
+    x_coords, y_coords = shift_and_rotate_waypoints(pose, waypoints, POINTS_TO_FIT)
+    coefficients = np.polyfit(x_coords, y_coords, 3)
+    distance = np.polyval(coefficients, 2.0)
```

Comment:
should we not use fit_polynomial defined above?

Reply:
we use it.. We get the coefficients and use them.

Reply:
Oh no.. That is another function that may be deleted. I see what you mean

Reply:
In this function I want to select only few waypoints...

Reply:
The other function just gets waypoints. Here I transform the coordinate system, select few and then fit

Reply:
ok. just wanted to double check we are not duplicating code